### ISSUE TITLE
Strange behaviors using interface types and extensions to add/replace fields

### ISSUE THREAD

**Issue by reinux** [ChilliCream/hotchocolate#4407]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

I'm trying to extend a type (call it `User`) which has a derived type (call it `Robot`). `User` has a field which returns an EF `IQueryable` of several child objects (call them `Message`), which also has a derived type (call it `QuitMessage`).

Several problems arise:

- Just exposing `Messages` as a field and trying to drill down into it using projections, it only returns objects of the base type, even if I have both the base and derived types registered using `AddObjectType` and `AddInterfaceType`.
- So I tried replacing it with a type extension, using `BindMember(nameof(User.Messages))]` causes another error message: 1. The field `messages` must be implement by object type `Robot`. (HotChocolate.Types.ObjectType<Robot>).
- Two more problems arise if I don't use `[BindMember]`:
  -  `Robot` will have a `messages` field, but not anything the base `User` type.
  - I want to hide the original `messages` field, because it only returns base type objects -- which I can't do without `[BindMember]`.
- In fact, `[BindMember]` seems to be being ignored completely; it just uses the method name to decide what to call the object.

### Steps to reproduce

Unfortunately, I haven't been able to start a new project in .NET 6 (mentioned [here](https://github.com/ChilliCream/hotchocolate/issues/1188#issuecomment-966045223)), so I'm not 100% sure this will behave the same way.

```
  class Message
  {
    public string Text { get; set; }
  }

  class QuitMessage : Message
  {
    public string Reason { get; set; }
  }

  class Robot : User
  {
    public string Bleep { get; set; }
  }

  class User
  {
    public string Name { get; set; }
    public IEnumerable<Message> GetMessages()
    {
      throw new NotImplementedException();
    }
  }

  class QueryType
  {
    public IEnumerable<Message> GetMessages()
    {
      throw new NotImplementedException();
    }
  }

  [ExtendObjectType(typeof(User))]
  public class UserExtensions
  {
    [UseDbContext(typeof(DataContext))]
    [UseProjection]
    //[BindMember(nameof(User.Messages))]
    //[BindMember("AllMessagesTest)]
    public IQueryable<Messages> GetAllMessages(
      [ScopedService] DataContext dc,
      [Parent] User user)
      => dc.Messages.Where(e => e.UserId == user.Id);
  }

builder.Services.AddGraphQLServer()
  .AddQueryType<QueryType>()
  .AddTypeExtension<UserExtensions>();
```

### Product

Hot Chocolate

### Version

12.1.0

**Comment by PascalSenn** [comment]
> Just exposing Messages as a field and trying to drill down into it using projections, it only returns objects of the base type, even if I have both the base and derived types registered using AddObjectType and AddInterfaceType.

Can you share your GraphQL Query? 

>So I tried replacing it with a type extension, using BindMember(nameof(User.Messages))] causes another error message: 1. The field messages must be implement by object type Robot. (HotChocolate.Types.ObjectType).
Two more problems arise if I don't use [BindMember]:
Robot will have a messages field, but not anything the base User type.
I want to hide the original messages field, because it only returns base type objects -- which I can't do without [BindMember].
In fact, [BindMember] seems to be being ignored completely; it just uses the method name to decide what to call the object.

In your example you bind the member `Messages`. 
isn't the member called `GetMessages` ?

**Comment by reinux** [comment]
Yes, sorry, that should be `Messages`, since it's an EF navigation property.

Unfortunately, I'm not able to isolate a repro right now, as I'm having trouble starting a new project in .NET 6 and I don't quite understand the error it's giving me ("Unable to infer or resolve a schema type from the type reference Output: QueryType.").

I'm sure I can get the repro working eventually. I'll report back when I get the chance to do that.

**Comment by PascalSenn** [comment]
can you publish that project on github? then i can have a look.

**Comment by reinux** [comment]
https://github.com/reinux/HotChocolateTest

**Comment by PascalSenn** [comment]
@reinux  It's because you classes are internal. Just add the `public` access modifier to the classes and it should work. 
I will have to have a look if we could also allow internal classes, but there could be a restriction somewhere.

**Comment by reinux** [comment]
Oh... Duh 😁

But yeah, would have been nice if the error message were clearer.

I'll continue on with this to see if I can reproduce the other issue.
